fastplotlib/layouts/_gridplot.py

```diff
@@ -110,127 +117,177 @@ def __init__(
+            types_flat = list(chain(*controller_types))
+            # str controller_type or pygfx instances
+            valid_str = list(valid_controller_types.keys()) + ["default"]
+            valid_instances = tuple(valid_controller_types.values())
+
+            # make sure each controller type is valid
+            for controller_type in types_flat:
+                if controller_type is None:
+                    continue
+
+                if (controller_type not in valid_str) and (
+                    not isinstance(controller_type, valid_instances)
+                ):
+                    raise ValueError(
+                        f"You have passed an invalid controller type, valid controller_types arguments are:\n"
+                        f"{valid_str} or instances of {[c.__name__ for c in valid_instances]}"
+                    )
 
-            cont_type = cont_type[0]
+            controller_types: np.ndarray[pygfx.Controller] = np.asarray(
+                controller_types
+            ).reshape(self.shape)
 
-            # get all the cameras that use this controller
-            cams = self._cameras[controller_ids == cid].ravel()
+            # make the real controllers for each subplot
+            subplot_controllers = np.empty(shape=self.shape, dtype=object)
+            for cid in np.unique(controller_ids):
+                cont_type = controller_types[controller_ids == cid]
+                if np.unique(cont_type).size > 1:
+                    raise ValueError(
+                        "Multiple controller types have been assigned to the same controller id. "
+                        "All controllers with the same id must use the same type of controller."
+                    )
 
-            if cont_type == "default":
-                # hacky fix for now because of how `create_controller()` works
-                cont_type = None
-            _controller = create_controller(controller_type=cont_type, camera=cams[0])
+                cont_type = cont_type[0]
 
-            self._controllers[controller_ids == cid] = _controller
+                # get all the cameras that use this controller
+                cams = subplot_cameras[controller_ids == cid].ravel()
 
-            # add the other cameras that go with this controller
-            if cams.size > 1:
-                for cam in cams[1:]:
-                    _controller.add_camera(cam)
+                if cont_type == "default":
+                    # hacky fix for now because of how `create_controller()` works
+                    cont_type = None
+                _controller = create_controller(
+                    controller_type=cont_type, camera=cams[0]
+                )
+
+                subplot_controllers[controller_ids == cid] = _controller
+
+                # add the other cameras that go with this controller
+                if cams.size > 1:
+                    for cam in cams[1:]:
+                        _controller.add_camera(cam)
```

Comment:
This entire block isn't new, it's now in the `else` block if `controllers is None`